rdrf/rdrf/api_views.py

```diff
@@ -0,0 +1,258 @@
+import pycountry
+
+from django.db.models import Q
+from rest_framework import generics
+from rest_framework import mixins
+from rest_framework import viewsets
+from rest_framework import serializers
+from rest_framework import status
+from rest_framework.exceptions import APIException
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.reverse import reverse
+from rest_framework.views import APIView
+
+from registry.genetic.models import Gene, Laboratory
+from registry.patients.models import Patient, Registry
+from registry.groups.models import CustomUser, WorkingGroup
+from serializers import PatientSerializer, RegistrySerializer, WorkingGroupSerializer, CustomUserSerializer
+
+
+import logging
+logger = logging.getLogger('registry_log')
+
+
+class BadRequestError(APIException):
+    status_code = status.HTTP_400_BAD_REQUEST
+
+
+class RegistryDetail(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Registry.objects.all()
+    serializer_class = RegistrySerializer
+    lookup_field = 'code'
+
+
+class RegistryList(generics.ListCreateAPIView):
+    queryset = Registry.objects.all()
+    serializer_class = RegistrySerializer
+
+
+class PatientDetail(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Patient.objects.all()
+    serializer_class = PatientSerializer
+
+    def _get_registry_by_code(self, registry_code):
+        try:
+            return Registry.objects.get(code=registry_code)
+        except Registry.DoesNotExist:
+            raise BadRequestError("Invalid registry code '%s'" % registry_code)
+
+    def check_object_permissions(self, request, patient):
+        """We're always filtering the patients by the registry code form the url and the user's working groups"""
+        super(PatientDetail, self).check_object_permissions(request, patient)
+        registry_code = self.kwargs.get('registry_code')
+        registry = self._get_registry_by_code(registry_code)
+        if registry not in patient.rdrf_registry.all():
+            self.permission_denied(request, message='Patient not available in requested registry')
+        if request.user.is_superuser:
+            return
+        if registry not in request.user.registry.all():
+            self.permission_denied(request, message='Not allowed to get Patients from this Registry')
+
+        working_groups = patient.working_groups.all()
+        if not any(map(lambda wg: wg in working_groups, request.user.working_groups.all())):
```

Comment:
What about `not patient.working_groups.filter(id__in=request.user.working_groups.all()).exists()` instead?